README.md
# Lazy Load Placeholders plugin for Craft CMS 3.x

Create blurry images from original images + lazy load functionality

#### Admin Setup
- Check assets folder in ```Settings > Lazy Load Placeholders```
- Click ```Generate Placehlders Now``` if you want generate plceholders 

#### Template Setup
- For get placeholder image add ```LLPlaceholder``` filter to asset Ex: ```{{ entry.image.one()|LLPlaceholder }}```
- If you want get base64 url add ```LLPlaceholder(true)``` filter to asset Ex: ```{{ entry.image.one()|LLPlaceholder(true) }}```

#### Lazy Load Setup
1. Lazy Load Images
    - Add class ```lazy-image```
    - Add placeholder asset to ```src``` ``` {{ entry.image.one()|LLPlaceholder }} ``` or ``` {{ entry.image.one()|LLPlaceholder(true) }} ```
    - Add normal asset url to ```image-src``` ```{{ entry.image.one().getUrl() }}```
    - EX: ```<img image-src="{{ entry.image.one().getUrl() }}" src="{{ entry.image.one()|LLPlaceholder }}" class="lazy-image">```

2. Lazy Load Background 
    - Add class ```lazy-background```
    - Add placeholder asset to ```style="background-image:url({{ entry.image.one()|LLPlaceholder }})" ``` or ``` {{ entry.image.one()|LLPlaceholder(true) }} ```
    - Add normal asset url to ```background-src="{{ entry.image.one().getUrl() }}"```
    - EX: ```<div class="lazy-background" background-src="{{ entry.image.one().getUrl() }}" style="background-image: url({{ entry.image.one()|LLPlaceholder }});"></div>```

3. Lazy Load Scripts
    - Add ```{{ loadLazyLoadScripts() }}``` in footer, for load lazy load scripts


CHANGELOG.md
# Lazy Load Placeholders Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 - 2019-04-12
### Added
- Initial release


src/models/Settings.php
<?php
/**
 * Google  reCAPTCHA v3 plugin for Craft CMS 3.x
 *
 * reCAPTCHA v3 returns a score for each request without user friction
 *
 * @link      https://github.com/Andr3y9603
 * @copyright Copyright (c) 2019 Ghiorghiu Andrei
 */

namespace ags\lazyloadplaceholders\models;

use ags\lazyloadplaceholders\LazyLoadPlaceholders;
use Craft;
use craft\base\Model;

/**
 * @author    Ghiorghiu Andrei
 * @package   GoogleRecaptchaV3
 * @since     1.0.0
 */
class Settings extends Model {
	// Public Properties
	// =========================================================================

	/**
	 * @var string
	 */
	public $volumes = [];
	public $volumesPaths = '';

	// Public Methods
	// =========================================================================

	/**
	 * @inheritdoc
	 */
	public function rules() {
		return [
		];
	}
}


src/LazyLoadPlaceholders.php
<?php
/**
 * Lazy Load Placeholders plugin for Craft CMS 3.x
 *
 * Create blurry images from original images + lazy load functionality
 *
 * @link      https://github.com/Andr3y9603
 * @copyright Copyright (c) 2019 Ghiorghiu Andrei
 */

namespace ags\lazyloadplaceholders;

use ags\lazyloadplaceholders\models\Settings;
use ags\lazyloadplaceholders\services\Functions as FunctionsService;
use ags\lazyloadplaceholders\twigextensions\LazyLoadPlaceholdersTwigExtension;
use Craft;
use craft\base\Plugin;
use craft\events\ElementEvent;
use craft\events\PluginEvent;
use craft\services\Elements;
use craft\services\Path;
use craft\services\Plugins;
use yii\base\Event;

/**
 * Class LazyLoadPlaceholders
 *
 * @author    Ghiorghiu Andrei
 * @package   LazyLoadPlaceholders
 * @since     1.0.0
 *
 * @property  FunctionsService $functions
 */
class LazyLoadPlaceholders extends Plugin {
	// Static Properties
	// =========================================================================

	/**
	 * @var LazyLoadPlaceholders
	 */
	public static $plugin;

	public $hasCpSettings = true;

	// Public Properties
	// =========================================================================

	/**
	 * @var string
	 */
	public $schemaVersion = '1.0.0';

	// Public Methods
	// =========================================================================

	/**
	 * @inheritdoc
	 */
	public function init() {
		parent::init();
		self::$plugin = $this;

		Craft::$app->view->registerTwigExtension(new LazyLoadPlaceholdersTwigExtension());

		Event::on(
			Plugins::class,
			Plugins::EVENT_AFTER_INSTALL_PLUGIN,
			function (PluginEvent $event) {
				if ($event->plugin === $this) {
				}
			}
		);

		Event::on(
			Elements::class,
			Elements::EVENT_AFTER_SAVE_ELEMENT,
			function (ElementEvent $event) {

				if (!$event->element instanceof craft\elements\Asset) {
					return $event;
				}

				$volumeId = $event->element->volumeId;
				$volumesPaths = LazyLoadPlaceholders::$plugin->getSettings()->volumesPaths;
				$volumes = LazyLoadPlaceholders::$plugin->getSettings()->volumes;
				$filename = $event->element->filename;

				if (!in_array($volumeId, $volumes)) {
					return false;
				}

				if (strlen(trim($volumesPaths[$volumeId])) === 0) {
					return false;
				}

				$originalImagePath = $volumesPaths[$volumeId];

				$path = Craft::getAlias('@webroot');
				$placeholderVolume = $path . '/placeholders';
				$filePath = $placeholderVolume . '/' . $filename;
				$rootPath = Craft::getAlias('@web');

				if (file_exists($filePath)) {
					return true;
				}

				if (!file_exists($placeholderVolume)) {
					mkdir($placeholderVolume);
				}

				$file = $originalImagePath . '/' . $filename;

				$type = exif_imagetype($file);

				if ($type === IMAGETYPE_JPEG) {
					$image = imagecreatefromjpeg($file);
				}
				if ($type === IMAGETYPE_PNG) {
					$image = imagecreatefrompng($file);
				}

				list($w, $h) = getimagesize($file);

				$size = array('sm' => array('w' => intval($w / 5), 'h' => intval($h / 5)),
					'md' => array('w' => intval($w / 2), 'h' => intval($h / 2)),
				);

				$sm = imagecreatetruecolor($size['sm']['w'], $size['sm']['h']);
				imagecopyresampled($sm, $image, 0, 0, 0, 0, $size['sm']['w'], $size['sm']['h'], $w, $h);

				for ($x = 1; $x <= 40; $x++) {
					imagefilter($sm, IMG_FILTER_GAUSSIAN_BLUR, 999);
				}

				imagefilter($sm, IMG_FILTER_SMOOTH, 99);
				imagefilter($sm, IMG_FILTER_BRIGHTNESS, 10);

				$md = imagecreatetruecolor($size['md']['w'], $size['md']['h']);
				imagecopyresampled($md, $sm, 0, 0, 0, 0, $size['md']['w'], $size['md']['h'], $size['sm']['w'], $size['sm']['h']);
				imagedestroy($sm);

				for ($x = 1; $x <= 25; $x++) {
					imagefilter($md, IMG_FILTER_GAUSSIAN_BLUR, 999);
				}

				imagefilter($md, IMG_FILTER_SMOOTH, 99);
				imagefilter($md, IMG_FILTER_BRIGHTNESS, 10);

				imagecopyresampled($image, $md, 0, 0, 0, 0, $w, $h, $size['md']['w'], $size['md']['h']);
				imagedestroy($md);

				imagejpeg($image, $placeholderVolume . '/' . $filename);

				imagedestroy($image);

			}
		);

		Craft::info(
			Craft::t(
				'lazy-load-placeholders',
				'{name} plugin loaded',
				['name' => $this->name]
			),
			__METHOD__
		);
	}

	// Protected Methods
	// =========================================================================

	/**
	 * @inheritdoc
	 */
	protected function createSettingsModel() {
		return new Settings();
	}

	/**
	 * @inheritdoc
	 */
	protected function settingsHtml(): string{
		$volumes = Craft::$app->getVolumes()->publicVolumes;

		$asda = Craft::getAlias('@web');

		$volumesMap = array_map(function ($item) {
			preg_match('/@[a-zA-Z]*/', $item->path, $matches, PREG_UNMATCHED_AS_NULL);

			if (count($matches) !== 0) {
				$path = Craft::getAlias($matches[0]);
				$path = preg_replace('/@[a-zA-Z]*/', $path, $item->path);
			} else {
				$path = Craft::getAlias('@webroot');
				$path .= '/' . $item->path;
			}

			return [
				'name' => $item->name,
				'path' => $path,
				'handle' => $item->handle,
				'id' => $item->id,
				'uid' => $item->uid,
			];
		}, $volumes);

		return Craft::$app->view->renderTemplate(
			'lazy-load-placeholders/settings',
			[
				'settings' => $this->getSettings(),
				'volumes' => $volumesMap,
			]
		);
	}

}


src/controllers/PlaceholdersController.php
<?php

namespace ags\lazyloadplaceholders\controllers;

use ags\lazyloadplaceholders\LazyLoadPlaceholders;
use Craft;
use craft\web\Controller;

class PlaceholdersController extends Controller {

	// Protected Properties
	// =========================================================================

	/**
	 * @var    bool|array Allows anonymous access to this controller's actions.
	 *         The actions must be in 'kebab-case'
	 * @access protected
	 */
	protected $allowAnonymous = ['index'];

	public function actionIndex() {

		$response = 'da';
		return json_encode(['da' => 123, 'nu' => 33]);

	}

	public function actionGenerateAll() {

		$volumeId = $_GET['volumeId'];
		$volumesPaths = LazyLoadPlaceholders::$plugin->getSettings()->volumesPaths;
		$volumes = LazyLoadPlaceholders::$plugin->getSettings()->volumes;
		$filename = $_GET['filename'];

		if (!in_array($volumeId, $volumes)) {
			return false;
		}

		if (strlen(trim($volumesPaths[$volumeId])) === 0) {
			return false;
		}

		$originalImagePath = $volumesPaths[$volumeId];

		$path = Craft::getAlias('@webroot');
		$placeholderVolume = $path . '/placeholders';
		$filePath = $placeholderVolume . '/' . $filename;
		$rootPath = Craft::getAlias('@web');

		if (file_exists($filePath)) {
			return true;
		}

		if (!file_exists($placeholderVolume)) {
			mkdir($placeholderVolume);
		}

		$file = $originalImagePath . '/' . $filename;

		$type = exif_imagetype($file);

		if ($type === IMAGETYPE_JPEG) {
			$image = imagecreatefromjpeg($file);
		}
		if ($type === IMAGETYPE_PNG) {
			$image = imagecreatefrompng($file);
		}

		list($w, $h) = getimagesize($file);

		$size = array('sm' => array('w' => intval($w / 5), 'h' => intval($h / 5)),
			'md' => array('w' => intval($w / 2), 'h' => intval($h / 2)),
		);

		$sm = imagecreatetruecolor($size['sm']['w'], $size['sm']['h']);
		imagecopyresampled($sm, $image, 0, 0, 0, 0, $size['sm']['w'], $size['sm']['h'], $w, $h);

		for ($x = 1; $x <= 40; $x++) {
			imagefilter($sm, IMG_FILTER_GAUSSIAN_BLUR, 999);
		}

		imagefilter($sm, IMG_FILTER_SMOOTH, 99);
		imagefilter($sm, IMG_FILTER_BRIGHTNESS, 10);

		$md = imagecreatetruecolor($size['md']['w'], $size['md']['h']);
		imagecopyresampled($md, $sm, 0, 0, 0, 0, $size['md']['w'], $size['md']['h'], $size['sm']['w'], $size['sm']['h']);
		imagedestroy($sm);

		for ($x = 1; $x <= 25; $x++) {
			imagefilter($md, IMG_FILTER_GAUSSIAN_BLUR, 999);
		}

		imagefilter($md, IMG_FILTER_SMOOTH, 99);
		imagefilter($md, IMG_FILTER_BRIGHTNESS, 10);

		imagecopyresampled($image, $md, 0, 0, 0, 0, $w, $h, $size['md']['w'], $size['md']['h']);
		imagedestroy($md);

		imagejpeg($image, $placeholderVolume . '/' . $filename);

		imagedestroy($image);

		return true;
	}

	public function actionGetFiles() {
		$volumesPaths = LazyLoadPlaceholders::$plugin->getSettings()->volumesPaths;
		$volumes = LazyLoadPlaceholders::$plugin->getSettings()->volumes;

		$volumes = array_filter($volumes);

		$groupNr = 5;
		$nr = 0;

		$filesGroups = [];
		for ($i = 0; $i < count($volumes); $i++) {
			$volumeId = $volumes[$i];
			$path = $volumesPaths[$volumeId];

			$files = array_filter(scandir($path), function ($v, $k) {
				return $v !== '.' && $v !== '..' && $v !== '.DS_Store';
			}, ARRAY_FILTER_USE_BOTH);

			sort($files);

			for ($j = 0; $j < count($files); $j++) {
				$file = $files[$j];

				if (empty($filesGroups[$nr])) {
					$filesGroups[$nr] = [];
				}

				$filesGroups[$nr][] = [
					'filename' => $file,
					'volumeId' => $volumeId,
				];

				if (($j + 1) % $groupNr === 0) {
					$nr++;
				}

			}

			return json_encode($filesGroups);
		}

	}
}

src/twigextensions/LazyLoadPlaceholdersTwigExtension.php
<?php
/**
 * Lazy Load Placeholders plugin for Craft CMS 3.x
 *
 * Create blurry images from original images + lazy load functionality
 *
 * @link      https://github.com/Andr3y9603
 * @copyright Copyright (c) 2019 Ghiorghiu Andrei
 */

namespace ags\lazyloadplaceholders\twigextensions;

use ags\lazyloadplaceholders\LazyLoadPlaceholders;
use Craft;

/**
 * @author    Ghiorghiu Andrei
 * @package   LazyLoadPlaceholders
 * @since     1.0.0
 */
class LazyLoadPlaceholdersTwigExtension extends \Twig_Extension {
	// Public Methods
	// =========================================================================

	/**
	 * @inheritdoc
	 */
	public function getName() {
		return 'LazyLoadPlaceholders';
	}

	/**
	 * @inheritdoc
	 */
	public function getFilters() {
		return [
			new \Twig_SimpleFilter('LLPlaceholder', [$this, 'LLPlaceholder']),
		];
	}

	/**
	 * @inheritdoc
	 */
	public function getFunctions() {
		return [
			new \Twig_SimpleFunction('loadLazyLoadScripts', [$this, 'loadLazyLoadScripts']),
		];
	}

	/**
	 * @param null $text
	 *
	 * @return string
	 */
	public function LLPlaceholder($asset = null, $exportBase64 = false) {
		$volumeId = $asset->volumeId;
		$volumesPaths = LazyLoadPlaceholders::$plugin->getSettings()->volumesPaths;
		$volumes = LazyLoadPlaceholders::$plugin->getSettings()->volumes;
		$filename = $asset->filename;

		if (!in_array($volumeId, $volumes)) {
			return false;
		}

		if (strlen(trim($volumesPaths[$volumeId])) === 0) {
			return false;
		}

		$originalImagePath = $volumesPaths[$volumeId];

		if (!$exportBase64) {

			$path = Craft::getAlias('@webroot');
			$placeholderVolume = $path . '/placeholders';
			$filePath = $placeholderVolume . '/' . $filename;
			$rootPath = Craft::getAlias('@web');

			if (file_exists($filePath)) {
				return $rootPath . '/placeholders/' . $filename;
			}

			if (!file_exists($placeholderVolume)) {
				mkdir($placeholderVolume);
			}
		}

		if ($exportBase64) {
			ob_start();
		}

		$file = $originalImagePath . '/' . $filename;

		$type = exif_imagetype($file);

		if ($type === IMAGETYPE_JPEG) {
			$image = imagecreatefromjpeg($file);
		}
		if ($type === IMAGETYPE_PNG) {
			$image = imagecreatefrompng($file);
		}

		list($w, $h) = getimagesize($file);

		$size = array('sm' => array('w' => intval($w / 5), 'h' => intval($h / 5)),
			'md' => array('w' => intval($w / 2), 'h' => intval($h / 2)),
		);

		$sm = imagecreatetruecolor($size['sm']['w'], $size['sm']['h']);
		imagecopyresampled($sm, $image, 0, 0, 0, 0, $size['sm']['w'], $size['sm']['h'], $w, $h);

		for ($x = 1; $x <= 40; $x++) {
			imagefilter($sm, IMG_FILTER_GAUSSIAN_BLUR, 999);
		}

		imagefilter($sm, IMG_FILTER_SMOOTH, 99);
		imagefilter($sm, IMG_FILTER_BRIGHTNESS, 10);

		$md = imagecreatetruecolor($size['md']['w'], $size['md']['h']);
		imagecopyresampled($md, $sm, 0, 0, 0, 0, $size['md']['w'], $size['md']['h'], $size['sm']['w'], $size['sm']['h']);
		imagedestroy($sm);

		for ($x = 1; $x <= 25; $x++) {
			imagefilter($md, IMG_FILTER_GAUSSIAN_BLUR, 999);
		}

		imagefilter($md, IMG_FILTER_SMOOTH, 99);
		imagefilter($md, IMG_FILTER_BRIGHTNESS, 10);

		imagecopyresampled($image, $md, 0, 0, 0, 0, $w, $h, $size['md']['w'], $size['md']['h']);
		imagedestroy($md);

		if ($exportBase64) {
			imagejpeg($image);
			$src = ob_get_clean();

			imagedestroy($image);

			return 'data:image/jpeg;base64,' . base64_encode($src);
		} else {
			imagejpeg($image, $placeholderVolume . '/' . $filename);
		}

		imagedestroy($image);

		return $rootPath . '/placeholders/' . $filename;
	}

	public function loadLazyLoadScripts() {
		$file = file_get_contents(__DIR__ . '/../templates/scripts.twig');

		echo $file;
	}
}
